File: Windows_csharp/RoutingMessage.cs
/*
 * Title:  spider RoutingMessage.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Net;
using System.Runtime.InteropServices;

namespace spider
{
    public class RoutingMessage : Message
    {
        private uint pipeId;


        public RoutingMessage()
        {
            this.pipeId = 0;
            this.messageType = 'r';
            this.dataSize = 0;
            this.data = new byte[SPIDER_MESSAGE_DATA_MAX_SIZE];
        }

        public RoutingMessage(uint pipeId,
                              byte[] array)
        {
            this.pipeId = pipeId;
            try
            {
                this.messageType = (char)array[0];
                this.dataSize = IPAddress.NetworkToHostOrder(BitConverter.ToInt32(array, 136));

                if(this.dataSize > 0)
                {
                    this.data = new byte[SPIDER_MESSAGE_DATA_MAX_SIZE];
                }

                if(this.dataSize > 0 &&
                   this.dataSize <= SPIDER_MESSAGE_DATA_MAX_SIZE)
                {
                    for(int i = 0; i < this.dataSize; i++)
                    {
                        this.data[i] = array[144 + i];
                    }
                }else if(this.dataSize > SPIDER_MESSAGE_DATA_MAX_SIZE)
                {
#if DEBUGPRINT
                    Console.WriteLine("[-] routing message data size error: {0}",
                                      this.dataSize);
#endif
                    this.dataSize = SPIDER_MESSAGE_DATA_MAX_SIZE;
                    for(int i = 0; i < SPIDER_MESSAGE_DATA_MAX_SIZE; i++)
                    {
                        this.data[i] = array[144 + i];
                    }
                }
            }catch(Exception ex)
            {
                throw new Exception("", ex);
            }
        }

        ~RoutingMessage()
        {

        }

        public uint PipeId
        {
            get { return pipeId; }
            set { pipeId = value; }
        }

        public ushort NetworkToHostOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public ushort HostToNetworkOrderUShort(ushort value)
        {
            if(BitConverter.IsLittleEndian)
            {
                byte[] bytes = BitConverter.GetBytes(value);
                Array.Reverse(bytes);
                return BitConverter.ToUInt16(bytes, 0);
            }

            return value;
        }

        public override int CopyToBuffer(ref byte[] buffer)
        {
            int length = 0;

            buffer[0] = (byte)this.messageType;
            BitConverter.GetBytes(IPAddress.HostToNetworkOrder(this.dataSize)).CopyTo(buffer, 136);
            for(int i = 0; i < this.dataSize; i++)
            {
                buffer[144 + i] = this.data[i];
            }

            length = SpiderMessageHeader.SPIDER_MESSAGE_HEADER_SIZE
                   + this.dataSize;

            return length;
        }
    }
}

File: extra/Windows/routingmessage.cpp
/*
 * Title:  spider routingmessage.cpp (Windows)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "routingmessage.hpp"


namespace spider
{
    Routingmessage::Routingmessage()
    {
        this->pipe_id = 0;
        this->message_type = 'r';
        this->data_size = 0;
        this->data = (char *)calloc(SPIDER_MESSAGE_DATA_MAX_SIZE,
                                    sizeof(char));
    }

    Routingmessage::Routingmessage(uint32_t pipe_id,
                                   struct spider_message *spider_message)
    {
        this->pipe_id = pipe_id;
        this->message_type = spider_message->header.message_type;
        this->data_size = ntohl(spider_message->header.data_size);

        if(this->data_size > 0)
        {
            this->data = (char *)calloc(SPIDER_MESSAGE_DATA_MAX_SIZE,
                                        sizeof(char));
        }

        if(this->data_size > 0 &&
           this->data_size <= SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
            std::memcpy(this->data,
                        spider_message->data,
                        this->data_size);
        }else if(this->data_size > SPIDER_MESSAGE_DATA_MAX_SIZE)
        {
#ifdef DEBUGPRINT
            std::printf("[-] routing message data size error: %d\n",
                        this->data_size);
#endif
            this->data_size = SPIDER_MESSAGE_DATA_MAX_SIZE;
            std::memcpy(this->data,
                        spider_message->data,
                        SPIDER_MESSAGE_DATA_MAX_SIZE);
        }
    }

    Routingmessage::~Routingmessage()
    {
        if(this->data != NULL)
        {
            free(data);
        }
    }

    void Routingmessage::set_pipe_id(uint32_t pipe_id)
    {
        this->pipe_id = pipe_id;
    }

    uint32_t Routingmessage::get_pipe_id()
    {
        return pipe_id;
    }

    int32_t Routingmessage::copy_to_buffer(char *buffer)
    {
        int32_t length = 0;

        struct spider_message *spider_message = (struct spider_message *)buffer;
        spider_message->header.message_type = this->message_type;
        spider_message->header.data_size = htonl(this->data_size);
        std::memcpy(spider_message->data,
                    this->data,
                    this->data_size);

        length = sizeof(struct spider_message_header)
               + this->data_size;

        return length;
    }
}

